### ISSUE TITLE
kind not working with Github Codespaces

### ISSUE THREAD

**Issue by jeremysf** [kubernetes-sigs/kind#2412]
I have created a .devcontainers.json that points to an debian image. I'm then running docker-in-docker inside of the debian  container that VSCode creates. In the debian container, I'm creating a kind cluster on the dockerd running inside of the debian container.

All of this works when VScode is launching the debian image on my Linux box (itself an debian distrubution).

However, when I point Github Codespaces at the .devcontainers.json, and it creates the debian container, no DNS works inside of the kind control plane/node container running on the dockerd running inside of the debian container.

The /etc/resolv.conf inside of the kind control plane looks "OK" in that it points back at the IP address of the debian container. This is the same as what I see when the debian container is running on my Linux box.

The DNS inside of other docker containers that I launch on the dockerd inside of the debian container work fine when running in Codespaces. It's just the kind control plane container whose DNS does not work.

Any idea why the kind control plane container would have its DNS fail where sibling containers have DNS that works fine?

Thank you!

**Comment by BenTheElder** [comment]
kind is doing a few things different with DNS than launching a simple container

1. It uses a non default docker network ("kind") which should enable Docker's embedded DNS (which is not enabled on the "bridge" network specially but is on all others)
2. In order to make this work with nested containers we rewrite it from loopback to an alternate IP and update docker's injected iptables rules for the embedded DNS socket.

It sounds like this is not happening properly.

https://github.com/kubernetes-sigs/kind/blob/c78bd3f581a1da72f514c3c6f8ef59f530d13b7d/images/base/files/usr/local/bin/entrypoint#L329

The rest of the questions from the bug template would be helpful, or `docker info`, `docker network inspect kind` and the node container logs + the full resolv.conf. without access to this environment it's hard to say what's different about it.

**Comment by jeremysf** [comment]
OMG, so sorry to skip template steps. Serves me right for submitting after a late night of debugging.

```
$ docker info
Client:
 Debug Mode: false

Server:
 Containers: 3
  Running: 3
  Paused: 0
  Stopped: 0
 Images: 3
 Server Version: 19.03.8
 Storage Driver: overlay2
  Backing Filesystem: <unknown>
  Supports d_type: true
  Native Overlay Diff: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.4.0-1055-azure
 Operating System: Debian GNU/Linux 10 (buster) (containerized)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 7.775GiB
 Name: codespaces_c47c86
 ID: 5VLT:MX67:J35M:PZSQ:V45G:GPVV:A2QT:X56J:XN5D:HZC6:BY4B:63NI
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: codespacesdev
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine

WARNING: No swap limit support
```

and

```
$ docker network inspect kind
[
    {
        "Name": "kind",
        "Id": "217b66ecf0ad76cc8eedb912f44dcef81e8ad80052dd2434deb59743ef60561d",
        "Created": "2021-08-12T22:03:08.385432703Z",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": true,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "172.19.0.0/16",
                    "Gateway": "172.19.0.1"
                },
                {
                    "Subnet": "fc00:f853:ccd:e793::/64",
                    "Gateway": "fc00:f853:ccd:e793::1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {
            "99de7078c09d1612cbb9feaa378ef044bad8d08a535d61443400142bdd74cd1a": {
                "Name": "kind-registry",
                "EndpointID": "2cd8d564c000efe2f151f6bb7fc6307ff3e288a15e30fb35a72753f289854b64",
                "MacAddress": "02:42:ac:13:00:03",
                "IPv4Address": "172.19.0.3/16",
                "IPv6Address": "fc00:f853:ccd:e793::3/64"
            },
            "bb5ce1bb44411d263de64ff3ca42c18b93d9a065a3fefea76983da142007d4d9": {
                "Name": "layout-control-plane",
                "EndpointID": "61d56bf64530c2545348d1ab93848c65fc6f0c5dd240bb6f228ed128342d527f",
                "MacAddress": "02:42:ac:13:00:02",
                "IPv4Address": "172.19.0.2/16",
                "IPv6Address": "fc00:f853:ccd:e793::2/64"
            }
        },
        "Options": {
            "com.docker.network.bridge.enable_ip_masquerade": "true",
            "com.docker.network.driver.mtu": "1500"
        },
        "Labels": {}
    }
]
```

I'll fill out the rest of the template steps below.

**Comment by jeremysf** [comment]
Here's the resolv.conf from the parent docker container where DNS works:
```
$ cat /etc/resolv.conf 
# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)
#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
# 127.0.0.53 is the systemd-resolved stub resolver.
# run "systemd-resolve --status" to see details about the actual nameservers.

nameserver 127.0.0.53
search ogmvq4kotf1ufmr35pbc14cdnc.xx.internal.cloudapp.net
options timeout:1 attempts:5
```

and the resolv.conf from the kind control plane node where DNS does not work:
```
nameserver 172.18.0.1
nameserver 2001:4860:4860::8888
nameserver 2001:4860:4860::8844
options ndots:0
```

and the resolv.conf from a (sibling to the control plane) nginx container I created with `docker nginx run` where DNS also works:
```

nameserver 8.8.8.8
nameserver 8.8.4.4
```

**Comment by jeremysf** [comment]
Actually, one clue, I am running a local registry (as a sibling to the control plane) and it does not seem to have DNS either :/

Here's the script I use to create the kind cluster:

```
#!/bin/bash

HOST_IP=`hostname -I | awk '{ print $1 ; exit }'`

set -euo pipefail

# create registry container unless it already exists
reg_name='kind-registry'
reg_port='5000'
running="$(docker inspect -f '{{.State.Running}}' "${reg_name}" 2>/dev/null || true)"
if [ "${running}" != 'true' ]; then
  docker run \
    -d --restart=always -p "${reg_port}:5000" --name "${reg_name}" \
    registry:2
else
echo "local registry already running, skipping creating"
fi

# create a cluster with the local registry enabled in containerd if it's not already running
kind_running="$(docker inspect -f '{{.State.Running}}' "layout-control-plane" 2>/dev/null || true)"
if [ "${kind_running}" != 'true' ]; then
cat <<EOF | kind create cluster --name layout --config=-
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
networking:
  apiServerAddress: "${HOST_IP}"
nodes:
- role: control-plane
  extraMounts:
  - hostPath: /workspaces
    containerPath: /workspaces
  extraPortMappings:
  - containerPort: 30000
    hostPort: 9000
    # optional: set the bind address on the host
    # 0.0.0.0 is the current default
    #listenAddress: "127.0.0.1"
    # optional: set the protocol to one of TCP, UDP, SCTP.
    # TCP is the default
    protocol: TCP
containerdConfigPatches:
- |-
  [plugins."io.containerd.grpc.v1.cri".registry.mirrors."localhost:${reg_port}"]
    endpoint = ["http://${reg_name}:${reg_port}"]
EOF

# connect the registry to the cluster network
INPSECT_NETWORK=`docker network inspect kind`
if [[ "$INPSECT_NETWORK" == *"$reg_name"* ]]; then
echo "kind network contains ${reg_name} skipping adding"
else
docker network connect "kind" "${reg_name}"
fi

else
echo "kind cluster already exists, skipping creating"
kubectl config use-context kind-layout
fi
```

**Comment by jeremysf** [comment]
```
$ kind version
kind v0.11.1 go1.16.4 linux/amd64
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.0", GitCommit:"c2b5237ccd9c0f1d600d3072634ca66cefdf272f", GitTreeState:"clean", BuildDate:"2021-08-04T18:03:20Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-21T23:01:33Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}
$ docker info
Client:
 Debug Mode: false

Server:
 Containers: 4
  Running: 2
  Paused: 0
  Stopped: 2
 Images: 3
 Server Version: 19.03.8
 Storage Driver: overlay2
  Backing Filesystem: <unknown>
  Supports d_type: true
  Native Overlay Diff: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.4.0-1055-azure
 Operating System: Debian GNU/Linux 10 (buster) (containerized)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 7.775GiB
 Name: codespaces_c47c86
 ID: 5VLT:MX67:J35M:PZSQ:V45G:GPVV:A2QT:X56J:XN5D:HZC6:BY4B:63NI
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: codespacesdev
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine

WARNING: No swap limit support
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

**Comment by jeremysf** [comment]
Oh, and finally:

```
$ kind export logs --name=layout
```
[logs.tar.gz](https://github.com/kubernetes-sigs/kind/files/6984166/logs.tar.gz)

**Comment by BenTheElder** [comment]
Thank you! And I missed this part:

> Here's the resolv.conf from the parent docker container where DNS works:

Ah, so this is "kind in docker in docker"?

If you create the nginx or registry container on the "kind" network (or anything other than the bridge), what /etc/resolv.conf do you see? I see that we're getting IPv6 addresses in there on the node container which is unexpected, we're expecting docker to inject 127.0.0.11 which we can then manipulate.

**Comment by jeremysf** [comment]
OMG, no, thank you for helping me with this!

Yes kind in docker, in docker.

Also, I think I can relatively easily create a Github repository that you can yourself click to get an all in browser Github Codespaces instance with kind if that helps at all.

Although this all seems a bit contrived, I think being able to use kind in a Github Codespace is a pretty slick and simple way to develop and iterate on a Kubernetes based application.

```
$ docker run --network kind nginx
```

The resulting resolv.conf is:

```
cat /etc/resolv.conf 
nameserver 127.0.0.11
nameserver 2001:4860:4860::8888
nameserver 2001:4860:4860::8844
options ndots:0
```

And at that point, DNS does not work in the nginx container.

**Comment by BenTheElder** [comment]
> Also, I think I can relatively easily create a Github repository that you can yourself click to get an all in browser Github Codespaces instance with kind if that helps at all.
>
> Although this all seems a bit contrived, I think being able to use kind in a Github Codespace is a pretty slick and simple way to develop and iterate on a Kubernetes based application.

It does sound really cool to me but my initial impression of codespaces is that to run anything (versus just code editing) you need paid billing? Looking again now it seems I got that wrong, I'll have to take this for a spin at some point.

> And at that point, DNS does not work in the nginx container.

OK, so that's problematic, something is broken with the docker-in-docker environment, I would expect a normal container without any of our network magic we need for further nesting to just work.

We should probably next check if it's any of the options used to create the kind network vs any other non-default network, e.g. the opt-in to ipv6, or the masquerade setting.

**Comment by BenTheElder** [comment]
Actually: https://kind.sigs.k8s.io/docs/user/resources/#using-kind-in-ci this repo seems like a good candidate to setup a sample codespaces config

**Comment by jeremysf** [comment]
Cool, will submit a PR there after lunch!

**Comment by jeremysf** [comment]
Ok, so I changed from my own custom docker-in-docker container (which just invoked dockerd) to using one based on Microsoft's VSCode/Codespaces docker-in-docker container:

https://github.com/microsoft/vscode-dev-containers/tree/main/containers/docker-in-docker

I tried to figure out what is different between the two; but I can't spot it. In any event, Microsoft's works great with kind!

So for now, I think while it would be interesting to understand the failure case, it definitely seems to as you point out above, something weird about the dockerd configuration/environment.
